### ISSUE TITLE
Feature Request: drag-drop image attachment

### ISSUE THREAD

**Issue by hematinik** [penge/my-notes#220]
Hi
It would be great if you could drag-drop images into the document instead of current "upload and address" method. I'd suggest [Dropzone.js](https://www.dropzonejs.com/) for this purpose (Im currently using it in a project and works fine).

P.S
current method is just fine and I don't want to overwhelm you with tickets, so please take it as an Idea and feel free to skip it ;)

**Comment by penge** [comment]
Hi @hematinik 

Sounds great. Thank you for the suggestion.

Here's how it could work:

- Drag and Drop image to upload it to your Google Drive (Google Drive Sync will need to be enabled as images will be uploaded there; images will be saved under My Notes / images)
- Once the image is uploaded, it is inserted into the note
- Only you can see the image, image URL cannot be open by others (Need to be signed-in to you Google Account in Chrome, that's the top right icon with a photo)

What do you think?

Will try to create a prototype version.

**Comment by hematinik** [comment]
I think there should be an image upload queue when you are working offline. the image gets inserted with a local URL and gets replaced with the cloud version whenever the sync process completes uploading.

I'd like to participate but I'm not familiar with the project structure and where to start learning about chrome extensions. I really appreciate if you guide me somehow.

**Comment by penge** [comment]
That would be definitely nice to have. The problem with that is, local storage is limited to 5MB. That could be quickly reached if one starts to save images locally, let's say one image can easily be 2MB or so. This could be solved if we decide to use `unlimitedStorage` and require it before installing the extension (if required permission is added later and one has My Notes already installed, it will automatically ask him to approve the required permission). Even if this would be used to provide increased storage, it's not optimal, as there would be no way to link those images, as there is no local file system, images would have to be embedded into the note itself, and then at some point replaced with URLs once they are uploaded.

Due to the complexity and limitations above described, I would suggest to not do this part offline. That would mean if one is offline, images are not loaded but other things continue to work as now. I think we are not that often offline, and if internet connection would be lost for a moment, note editing would continue to work, with images to be the only exception. If My Notes would be let's say an Electron app, those limitations with storage would go away, but also the simple authentication against Google Drive and Chrome APIs we can use now. Electron supports a subset of Chrome APIs.

To learn about the extensions, please see these websites:
- https://developer.chrome.com/docs/extensions/mv3/getstarted/ (main takeaway is, manifest.json is the entrypoint in each extension)
- https://developer.chrome.com/docs/extensions/mv3/manifest/ (the list of all things `manifest.json` provides)
- https://developer.chrome.com/docs/extensions/reference/ (all available APIs, some of them work out of the box, or out of the box to some extend, some APIs will need additional permissions, these are requested in `manifest.json`)

Here are some articles you may also want to check out:
- https://dev.to/penge/chrome-extension-that-skips-youtube-ads-steps-how-to-create-it-3ibp
- https://dev.to/penge/learn-the-most-useful-chrome-apis-by-creating-block-site-chrome-extension-2de8

I hope this will help.

**Comment by penge** [comment]
Let's do it.

**Comment by hematinik** [comment]
that's great *_* 
I apologize for my late response, very useful information indeed! I started going through the list today but don't think I'll catch up with you in this particular feature.

there's a similar chrome extension which provides image drag and drop functionality: [Notes for Google Drive](https://chrome.google.com/webstore/detail/notes-for-google-drive/ndidogegapfaolpcebadjknkdlladffa?hl=en)

however the notes panel disappears when lost focus, so you have to Inspect the window (open up chrome dev tools) to keep it open for the drag n drop! (yes that's silly)
I thought it might be helpful to you like using the same logic or maybe giving you a starting point.

**Comment by penge** [comment]
Thanks a lot for suggesting **Notes for Google Drive**. I have looked over the reviews as I think it could be great source for inspiration.

The most requested feature there is **offline support**. That is quite a big deal and I am glad we have that in My Notes.

Another requested feature there is to see note in a tab (more space to edit). This is quite easy to do. Found the source code, at https://github.com/esheppard/notesfordrive. It doesn't seem maintained (2 commits in previous year, few in 2018, few in 2016, 2015, project created in **2014**).

One thing I like about Notes for Google Drive, last used text foreground and background color is remembered, to reuse it (this feature comes with Summernote). We could have that in My Notes.

What I really don't like, is that note name has to be as the first line in the note. It goes against how we work with files. Also leads to accidental file renames and problems anytime the first line in that note is edited.

From the chosen approach, we both use **Google Drive API** to save notes to Google Drive. Editor in My Notes is created from scratch as I found that to be a nice exercise. Notes for Google Drive uses [Summernote](https://github.com/summernote/summernote) (which is perfectly fine) that also brings additional dependecies: jQuery, Bootstrap, Summernote itself.

My Notes is a Nodejs project created in [Typescript](https://www.typescriptlang.org). Frontend is in [Preact](https://preactjs.com). Tests are in [Jest](https://jestjs.io). And it's bundled by awesome [esbuild](https://esbuild.github.io).

Other than that, I am glad **Notes for Google Drive** helps so many people. Unfortunately, it's not maintained. I know, it's quite lot of work.

**Comment by penge** [comment]
Regarding the **Drag & Drop** to upload images, here's what I find best.

You will Drag & Drop the image and immediately see the silhouette image in place, to have the idea of its dimensions.

Silhouette starts to fill from the left to the right, that will be the upload progress.

Once uploaded, image is in place.

What do you think?

**Comment by penge** [comment]
I think we will have a pretty nice and unique image upload.


https://user-images.githubusercontent.com/907255/115968308-923a2300-a537-11eb-9efb-e75818caf023.mov

**Comment by hematinik** [comment]
wow! this is awesome
the effect looks so slick. you are an artist @penge

**Comment by penge** [comment]
Thank you!

**Comment by penge** [comment]
Great news!

Adding a new metric – **Images** (showing the number of images in the active note):

<img width="569" alt="my_notes_images_count" src="https://user-images.githubusercontent.com/907255/116006135-3f7d6b80-a60a-11eb-86ad-72a9ee23ac3f.png">

Next, **_double-click_** on any image will copy its URL to the clipboard! _(Just thought that could be something really simple and handy.)_

---

Moving to the final step, that is uploading the images to **Google Drive**. Before getting there, I have finished the hard part (as in the video), as well first tried to upload the images to a free service – [imgur](https://imgur.com) – where I realized, it's actually pretty cool and could stay as a feature.

So after we will have image upload to Google Drive, we can start thinking about adding more image services, and even selecting the destination before the upload (Always ask), or use the one set in the Options (Google Drive, imgur, etc.).

**Comment by hematinik** [comment]
Regarding the double-click functionality, it does not sound very intuitive since we are working in a text editor and there's good chance that you already have something important in the clipboard.
it feels more natural to enlarge the image or open it in a new tab using double-click and in case you need the image's URL we can provide it in a contextual menu.

the metrics are cool as long as they are not being calculated as you go since It may cause input latency on slower devices. I also like the alternative image server Idea, google drive is somewhat slow on loading images IMO.

**Comment by penge** [comment]
^ Great feedback, thanks. So you know what, let's remove that one feature. At least it won't replace the clipboard accidentally.

Right-click on image to copy the URL is always there as option.

<img width="1312" alt="Screen Shot 2021-04-26 at 14 58 59" src="https://user-images.githubusercontent.com/907255/116086492-2921ef00-a6a0-11eb-954a-90d8e364ff88.png">

**Comment by penge** [comment]
### **Update:**

Google Drive image upload is working. Proceeding to testing.

**How it works:**

1. Drag & Drop the image into the note.
2. Image is instantly placed into the note as a skeleton (looks like the final image but semi-transparent).
3. Image is being uploaded immediately, image starts to reveal itself as semi-transparent goes away from the left to the right (the upload progress).
4. The image is nicely visible as it is now uploaded, you can continue editing the note.

\* Because image upload can be very quick, step 3 can look like never happened, the image is just uploaded instantly.

<br>

## QA

**1. Where are images uploaded?**

Images are uploaded to your personal Google Drive under the folder My Notes / assets. Folder "assets" is created automatically when needed.

**2. Are images stored privately?**

Yes. Only you can see the images. If you'd like to share the image(s) publicly or to someone specific, see Share button in Google Drive.

**3. Are images deleted from my Google Drive after I remove them from a note?**

No. Images continue to be stored in your Google Drive for future uses. This way images don't have to be uploaded again. If you'd like to delete the image permanently, go to Google Drive to delete the image.

**4. How can I insert the once uploaded image after I have removed it from a note?**

Go to your Google Drive, see the image you'd like to insert, right-click on the image and click on "Get link". Copy the link and paste to My Notes using the "Insert Image" button.
